Matrix multiplicative factor (z + z == 2 * z)

romainfd opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fixes https://github.com/Qiskit/qiskit-aqua/issues/1095
I have added a unittest for that, not changed any documentation (as the change is not changing any signature and better matches the expected behavior of `equals`) and not added any CHANGELOG as I couldn't find "the CHANGELOG file under Unreleased section".

### Details and comments

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-aqua?pullRequest=1111) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-aqua?pullRequest=1111) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-aqua?pullRequest=1111) it.</sub>

woodsp-ibm commented on qiskit/aqua/operators/primitive_ops/matrix_op.py:
```diff
@@ -95,10 +95,9 @@ def adjoint(self) -> OperatorBase:
             return False
-
-        return self.primitive == other.primitive
+        return self.coeff * self.primitive == other.coeff * other.primitive
```
I think the right solution to this is more complex than you have given coeff can include things like ParameterExpression. You can see an aspect of this in add().

Also, I think that if we cannot do the comparison we should be raising NotIimplemented rather than returning False such as the test for other being MatrixOp or not. I know this is from the original code, and not part of your change as such, but its something that we need to re-consider so I am just noting it here.

romainfd replied:
Thanks for your review!

Indeed! Thanks for the flag! I added basic expression comparison. Not really sure how to handle the issue for this type of parameter (like `"2 * y" * X == "y" * (2.X)`) as I couldn't find an easy way in `sympy` to get the possible factorisation between 2 expressions...

I understand your point and tried it but this caused the `reduce` function to fail on this line https://github.com/Qiskit/qiskit-aqua/blob/95d89596cb4670f05530aad2ff77724fe306118a/qiskit/aqua/operators/list_ops/summed_op.py#L97
Based on this, I think it actually makes sense to return `False` for not implemented comparisons.

woodsp-ibm replied:
I imagined that returning NotImplemented may take adjustments elsewhere - hence it was more of a note. Its something that needs further investigation across the board outside of the scope of this PR.

We ideally ought to have a unit test that goes through the path of Parameter checking where it returns False .

I think the logic you have now is ok, thanks for the contribution. I will note the the primitive can still be different types but the Operator from Terra quantum_info, one of the types, supports multiply so I think its all good. A mix of different primitive types will continue to return False, as indeed it did before (even if they may really represent the same thing).

romainfd replied:
I see! If you eventually decide to go with this `NotImplementedError` change and if needed, I can help with that :) 

I added a unit test checking `equals` returns `False` on different coefficient types (`expr * z != 2 * z`). Is it what you meant or you wanted a unit test for `expr1 * z != expr2 * z`?

Thank you for your reviews. You're welcome, my pleasure to help :) Yes indeed, I have tried to learn from your previous comment and looked at the types hints this time. From what I have seen, Operator implements `_multiply()` which is matched to `*` by its parent class `BaseOperator` so this part should indeed be fine. But when removing the type ignore comment, mypy was still failing because multiplications between `complex` and `list` are not supported. I copied the type fix from the `add` function just above the `equals` one https://github.com/Qiskit/qiskit-aqua/blob/95d89596cb4670f05530aad2ff77724fe306118a/qiskit/aqua/operators/primitive_ops/matrix_op.py#L84-L88

I also agree on your last point. Though this might not be the perfect behavior right now, I think this change should live in another issue/PR focused on converting between possible primitive types. I can also help on that if needed though I might need details/links on the expected behavior when comparing different representations.

woodsp-ibm replied:
Thanks, the unit test you added was indeed what I had in mind. 

The primitive type in this derived class MatrixOp is more constrained than the types of its base class. list as you see is converted an Operator so the type ignore ends up being needed for what mypy sees as the full set types which are in fact less here. Thanks for looking through that though and checking too.

romainfd commented:
Just noticed my PR was creating a useless `master` file... Sorry :/

romainfd commented:
@Cryoris,
Thanks for comment!
As suggested I updated `MatrixOp.add` to sum coeffs is the primitives are the same. I put the logic before the checks of the coeffs not being `ParameterExpression` as I've seen the `ParameterExpression` type implements `__add__` but can move it down if needed.
I added a unittest testing all your permutations. I don't know how you feel about several `asserts` in a single unittest and can reduce it, split it or do some aggregation before the assert but I think it's more readable that way and the error in case of failure will be easy enough to understand.
I was also wondering what you thought of updating `SummedOp.equals` with a block similar too https://github.com/Qiskit/qiskit-aqua/blob/95d89596cb4670f05530aad2ff77724fe306118a/qiskit/aqua/operators/list_ops/summed_op.py#L189-L195
but looping on the `oplist` and migrating the coeffs inside the primitives. Basically absorbing the coeffs one level lower by reducing all ops to coeffs=1 before comparing. I believe this would solve your comment and also provide that to other `Operator`s, am I wrong?

woodsp-ibm commented:
In regards of the build failure it was due to a new library version that came out today. Updating the branch should fix this by merging the latest changes into your branch.

Cryoris commented:
Yeah we could change the `SummedOp.equals` to absorb the coefficient of `MatrixOp`s into the primitive. But the matrix op here is special since its the only one where the coefficient can be absorbed into the primitive since the primitive is numerical. 
Or is there a use-case you have in mind that we are not covering with this change?

romainfd commented:
Thanks for the branch update and the details @woodsp-ibm 
@Cryoris, I had no use-case in mind and was just wondering / trying to grasp a bit more what each part does :)
